File: templates/python/v1/templatemethod/report_generators.py
from __future__ import annotations

from dataclasses import dataclass

SUMMARY_PREFIX = "Summary report: "
DETAIL_PREFIX = "Detailed report:"
DETAIL_HEADER = "[DETAIL]"
LINE_BREAK = "\n"
BULLET_PREFIX = "- "


@dataclass(slots=True)
class BaseReportGenerator:
    def generate(self, input_text: str | None) -> str:
        normalized = self.normalize(input_text)
        body = self.build_body(normalized)
        return self.format(body)

    def normalize(self, input_text: str | None) -> str:
        return (input_text or "").strip()

    def build_body(self, normalized: str) -> str:
        raise NotImplementedError

    def format(self, body: str) -> str:
        return body


@dataclass(slots=True)
class SummaryReportGenerator(BaseReportGenerator):
    def build_body(self, normalized: str) -> str:
        return f"{SUMMARY_PREFIX}{normalized}"


@dataclass(slots=True)
class DetailedReportGenerator(BaseReportGenerator):
    def build_body(self, normalized: str) -> str:
        return _format_bulleted_body(normalized)

    def format(self, body: str) -> str:
        return f"{DETAIL_HEADER}{LINE_BREAK}{body}"


def _format_bulleted_body(text: str) -> str:
    return f"{DETAIL_PREFIX}{LINE_BREAK}{BULLET_PREFIX}" + text.replace(
        LINE_BREAK, f"{LINE_BREAK}{BULLET_PREFIX}"
    )

File: templates/javascript/v1/templatemethod/reportGenerators.js
export class ReportGenerator {
  generate(input) {
    const normalized = this.normalize(input);
    const body = this.buildBody(normalized);
    return this.format(body);
  }

  normalize(input) {
    return (input ?? "").trim();
  }

  buildBody(_normalized) {
    throw new Error("Not implemented");
  }

  format(body) {
    return body;
  }
}

export class SummaryReportGenerator extends ReportGenerator {
  buildBody(normalized) {
    return `Summary report: ${normalized}`;
  }
}

export class DetailedReportGenerator extends ReportGenerator {
  buildBody(normalized) {
    return `Detailed report:\n- ${normalized.split("\n").join("\n- ")}`;
  }

  format(body) {
    return `[DETAIL]\n${body}`;
  }
}
